File: problems/5-Longest-Palindromic-Substring/solution.go
func longestPalindrome(s string) string {
	checkList := make([][]int, len(s))
	for i := 0; i < len(s); i++ {
		checkList[i] = make([]int, 2)
		checkList[i][0] = 0
		checkList[i][1] = 1
	}

	max := 1
	result := string(s[0])
	for i := 1; i < len(s); i++ {
		for j := 0; j < len(checkList[i-1]); j++ {
			if i-1-checkList[i-1][j] >= 0 && s[i-1-checkList[i-1][j]] == s[i] {
				checkList[i] = append(checkList[i], checkList[i-1][j]+2)
				if checkList[i-1][j]+2 > max {
					max = checkList[i-1][j] + 2
					result = s[i-1-checkList[i-1][j] : i+1]
				}
			}
		}
	}

	return result
}

File: problems/5-Longest-Palindromic-Substring/solution.java
class Solution {
    public String longestPalindrome(String s) {
        List<List<Integer>> checkList = new ArrayList<>();
        for (int i = 0; i < s.length(); i++) {
            checkList.add(new ArrayList<>());
            checkList.get(i).add(0);
            checkList.get(i).add(1);
        }

        int maxLen = 1;
        String maxStr = s.substring(0, 1);
        for (int i = 1; i < s.length(); i++) {
            for (int j = 0; j < checkList.get(i - 1).size(); j++) {
                if (i - 1 - checkList.get(i - 1).get(j) >= 0 && s.charAt(i - 1 - checkList.get(i - 1).get(j)) == s.charAt(i)) {
                    checkList.get(i).add(checkList.get(i - 1).get(j) + 2);

                    if (maxLen < checkList.get(i - 1).get(j) + 2) {
                        maxLen = checkList.get(i - 1).get(j) + 2;
                        maxStr = s.substring(i - 1 - checkList.get(i - 1).get(j), i + 1);
                    }
                }
            }
        }

        return maxStr;
    }
}

